### PR TITLE
Msg tracker - parsing

### PR DESCRIPTION
**Pull request by AlbertoSoutullo**
Small PR that fragments the new message tracking utility for the repository.

The data given by its methods will be used for in a graph representation to show the flow of the messages in the mesh.

Edit: Related to #22 #23

### CODE REVIEW

File: src/mesh_analysis/tracers/message_tracer.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+from abc import ABC, abstractmethod
+from typing import List
+
+
+# Project Imports
+
+
+class MessageTracer(ABC):
+
+    @abstractmethod
```
Comment (AlejandroCabeza) [inline src/mesh_analysis/tracers/message_tracer.py:12]: You don't need the `@abstractmethod` or the `pass` in this method.

File: src/mesh_analysis/tracers/waku_tracer.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        # - Different patterns (received, sent, dropped)
+        # - Once one pattern search is completed, stop search for it in the logs (ie: Announce Address)
+        self._patterns = [re.compile(
+            r'^(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+\+\d{2}:\d{2}) .* msg_hash=([a-fA-F0-9x]+) .* sender_peer_id=([A-Za-z0-9]+)$'),
+            re.compile(r'.* Announcing addresses .*\[([^]]+)\]$')]
+
+    def trace(self, parsed_logs: List) -> pd.DataFrame:
+        df = self._trace_message_in_logs(parsed_logs)
+
+        return df
```
Comment (AlejandroCabeza) [inline src/mesh_analysis/tracers/waku_tracer.py:24]: As much as I love splitting stuff into functions, this one is not proving any value, only an indirection.

File: src/mesh_analysis/tracers/waku_tracer.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    def _trace_message_in_logs(self, parsed_logs: List) -> pd.DataFrame:
+        parsed_logs = [log for log in parsed_logs if len(log[0]) > 0]
```
Comment (AlejandroCabeza) [inline src/mesh_analysis/tracers/waku_tracer.py:27]: Not that it matters a lot in this case, but as a general rule I'd recommend using generators over list comprehensions when possible.

File: src/mesh_analysis/readers/file_reader.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+        if files_result.is_ok():
+            parsed_logs = self._read_files(files_result.ok_value)
+            logger.info(f'Tracing {self._folder}')
+            df = self._tracer.trace(parsed_logs)
+        else:
+            logger.error(f'Could not read {self._folder}')
+            exit()
+
+        return df
```
Comment (AlejandroCabeza) [inline src/mesh_analysis/readers/file_reader.py:35]: Styling opinion: When one of the `if` arms makes an early exit, I generally favour doing that check first; but I'm fine as it is :)
```suggestion
        if not files_result.is_ok():
            logger.error(f'Could not read {self._folder}')
            exit()

        parsed_logs = self._read_files(files_result.ok_value)
        logger.info(f'Tracing {self._folder}')
        return self._tracer.trace(parsed_logs)
```

File: src/mesh_analysis/readers/file_reader.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    def _read_files(self, files: List) -> List:
+        # TODO: set this as a parameter?
+        num_processes = multiprocessing.cpu_count()
```
Comment (AlejandroCabeza) [inline src/mesh_analysis/readers/file_reader.py:39]: `num_processes` is unused.
Also, what happens if the `processes=4` is higher than `num_processes`?

File: src/mesh_analysis/readers/file_reader.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        with multiprocessing.Pool(processes=4) as pool:
+            parsed_logs = pool.map(self._read_file_patterns, files)
+
+        return parsed_logs
+
+    def _read_file_patterns(self, file: str) -> List:
+        results = [[] for p in self._tracer.patterns]
```
Comment (AlejandroCabeza) [inline src/mesh_analysis/readers/file_reader.py:46]: Alternative, just FYI
```suggestion
results = [[]] * self._tracer.patterns.len
````

File: src/mesh_analysis/tracers/waku_tracer.py (hunk 2 above)
Comment (AlbertoSoutullo) [reply src/mesh_analysis/tracers/waku_tracer.py:24]: I agree. This function will be enhanced, tho. This is not the complete final PR.

**Comment by AlbertoSoutullo** [comment]
Thx for the review!

**Review by AlejandroCabeza — APPROVED** [review]

**Pull request merged**